### DISCUSSION TITLE
用Python寫簡易的App

### DISCUSSION THREAD

**Discussion by HsuChen88** [1101-GS4719-C/Python-Community-GS4719-C#562] [General]
# 什麼是GUI?
圖形使用者介面(Graphical User Interface)，圖形使用者介面是一種人與計算機通訊的介面顯示格式，允許使用者使用滑鼠等輸入裝置操縱螢幕上的圖示或選單選項，以選擇命令、呼叫檔案、啟動程式或執行其它一些日常任務。

現今市面上不缺面向Python的GUI框架。**Tkinter**、**WxPython**、**Qt**和**Kivy**是幾個主要的軟體包。此外，還有大量簡化版GUI軟體包可「包裝」其中一個主要軟體包，包括**EasyGUI**、**PyGUI**和**Pyforms**。但對於大多數初學者來說，構建自定義GUI佈局依然很困難或不可能。

## PySimpleGUI
**PySimpleGUI** (<https://github.com/MikeTheWatchGuy/PySimpleGUI>)試圖通過為GUI提供一種超級簡單、易於理解且易於定製的界面來克服這些GUI挑戰。若使用PySimpleGUI，連許多複雜的GUI也用不了20行代碼。

# 安裝PySimpleGUI

### 通過python pip安裝
``` 
pip install PySimpleGUI
```

# 什麼是Elements?
Element是整個**PySimpleGUI**框架的核心，Element可以是一個Text、image、slider或在任何程式中看到任何類型的元素，白話來說，任何看的到的元素就叫Element，包括視窗；也有顯示視窗看不見的Element，而這些Element通常用於佈局(layouts)。

# Elements的使用說明
### [ELEMENT AND FUNCTION CALL REFERENCE](https://pysimplegui.readthedocs.io/en/latest/call%20reference/)

# 快速創建GUI

### [PySimpleGUI Cookbook](https://pysimplegui.readthedocs.io/en/latest/cookbook/)

如果你有簡單的佈局，應該用不了5分鐘，就能夠用PySimpleGUI創建自定義佈局，只要修改來自Cookbook的代碼。，只要修改來自Cookbook的代碼。

1.找到類似你想創建的GUI的一個GUI。
2.從Cookbook複製代碼。
3.粘貼到你的IDE，並運行。

原文網址：https://kknews.cc/code/e4n3nmn.html

# 自己寫一個App
在這邊我就不贅述了，但推薦大家可以去看下方的影片。:sweat_smile:

# PySimpleGUI的缺點

### 外觀的限制
雖然**PySimpleGUI**有一些內建的選項可以製作GUI，也可以自己創建樣式，但通常不會花時間去創建樣式，因為創建的樣式也無法看起來和商業應用的一樣好，所以在樣式方面還是相當有限的。

### 規模的限制
**PySimpleGUI**是為小型 apps 設計的，如果想讓專案涵蓋更複雜的內容和華麗的格式，使用**PySimpleGUI**會讓自己遇到許多困境。

# 簡易實作範例影片
## [Creating 10 apps in Python with PySimpleGUI](https://www.youtube.com/watch?v=QeMaWQZllhg)
1. Unit Converter
2. Calculator
3. Stopwatch
4. Text Editor
5. Snake
6. Graph App (MatPlotLib)
7. Image Editor (Pillow)
8. Youtube App (Pytube)
9. Weather App (BeautifulSoup4)
10. Face Detector (OpenCV)
---
以上為我的新發現，也是第一次用Markdown排版，請見諒。
對任何事有想法或建議的，都歡迎在下方留言:stuck_out_tongue:。

**Comment by PoooYen** [comment]
這已經有一點進階了，相信對初學者學習有巨大的加分效果

**Comment by jason79461385** [comment]
Matlab也有相似的東西呢，改天再來研究看看
